[scripts] mock objects (#13)
SymfonyCasts/testing: sfcasts/ep1/mocking-mock-objects.md

# Mocking: Mock Objects

Our tests are passing, the dino's are wandering, and life is great! But... let's
think about this for a second. In `GithubService`, when we test `getHealthReport()`,
we're able to control the `$response` that we get back from `request()` by using
a stub. But we also need to ensure that the service is only calling GitHub one time,
*and* it's using the right HTTP method with the correct URL. How do we do that?

In `GithubServiceTest` where we configure the `$mockHttpClient`, add `->expects()`,

> and pass in the `once()` `InvocationRule` needed for this method.

Reviewer: That got a little too technical - I'm going to reword slightly.

Over in the terminal, run our tests...

```terminal
./vendor/bin/phpunit
```

And... Awesome! We've just added an assertion to our mock client that requires the
`request` method to be called *exactly* one time. Let's take it a step further and
add `->with()` to the `$mockHttpClient`. For the arguments, pass in `GET` and the

> URL for GitHub: `https://api.github.com/repos/SymfonyCasts/dino-park/`.

Author: Hmmm... im not going to say the actual url in the video..  Think this should be
> For the arguments, pass in `GET` and then I'll copy the URL for GitHub...

Reviewer: > For the arguments, pass in GET and then I'll paste the URL we need to the GitHub API.

Let's try the tests again...

```terminal-silent
./vendor/bin/phpunit
```

And... Huh! We have 2 failures:

> Failed asserting that two strings are equal

Hmm... Ah Ha! My copy and paste skills, are a bit weak... I missed `/issue`
at the end of the URL. Go ahead and add that.

> I think that should do the trick, let's find out...

Author: > ....at the end of the URL. Add that and I think that should do the trick. Let's find out...

Reviewer: agreed!

```terminal-silent
./vendor/bin/phpunit
```

> Umm... Yes! We're green all day... But best of all, we are 100% certain that we
> are using the correct URL and HTTP Method when we call GitHub.

Author: ```suggestion
Umm... Yes! We're green all day... But best of all, the tests confirm we
are using the correct URL and HTTP Method when we call GitHub.
```

*is* being called in our service. But, we already have other tests in place that
would fail if we *didn't* call that method. Let's just pick and choose our battles,
we can always add that later if we need to.

Now that `GithubService` is fully tested, let's keep the dashboard updated in realtime. On
`MainController::index()`, add argument: `GithubService $github`. Anytime someone
visit's the dashboard, Symfony will automatically find our service and inject it
into this method.

Right below the `$dinos` array, add a `foreach()` loop and for `$dinos as $dino`,
say `$dino->setHealth()` by calling `$github->getHealthReport()` using
`$dino->getName()`.

To the browser and refresh...

And... What!

> getDinoStatusFromLabels must be HealthStatus, null returned

Hmm... What is going on here. I can't really tell *why* we're getting this error.
In a future tutorial, we'll actually be able to add a functional test for this, 
but for now it looks like one of our dino's has a status label that we don't 
know about. Let's take a peak back at the issues on GitHub and... HA! "Dennis"
is causing problems yet again. Apparently he's a bit hungry...

> In our `HealthStatus` enum, we *don't* have a case for `Hungry` status labels. 
> Go figure. Next up, we'll see how we can provide a better exception when this 
> happens since its best to leave the investigative work to GenLab...

Author: I don't like this now that im reading it again.. 

Something like: 
```suggestion
In our `HealthStatus` enum, we *don't* have a case for `Hungry` status labels. 
Go figure. Next up, we'll come up with a better exception message in case this 
happens again, since its best to leave the investigative work to GenLab...
```
